detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/internal/RuleSetConfigProperty.kt

```diff
@@ -0,0 +1,29 @@
+package io.gitlab.arturbosch.detekt.api.internal
+
+import io.gitlab.arturbosch.detekt.api.Config
+import kotlin.properties.ReadOnlyProperty
+import kotlin.reflect.KProperty
+
+fun <T : Any> ruleSetConfig(defaultValue: T): ReadOnlyProperty<Any?, RuleSetConfigProperty<T>> =
+    RuleSetConfigPropertyDelegate(defaultValue)
+
+private class RuleSetConfigPropertyDelegate<T : Any>(
+    val defaultValue: T
+) : ReadOnlyProperty<Any?, RuleSetConfigProperty<T>> {
+
+    @Volatile private var _value: RuleSetConfigProperty<T>? = null
+
+    override fun getValue(thisRef: Any?, property: KProperty<*>): RuleSetConfigProperty<T> =
+        if (_value != null) {
+            _value as RuleSetConfigProperty<T>
+        } else {
+            RuleSetConfigProperty(property.name, defaultValue).also {
+                _value = it
+            }
+        }
+}
+
+class RuleSetConfigProperty<T : Any>(val key: String, val defaultValue: T)
+
+fun <T : Any> Config.value(ruleSetConfigProperty: RuleSetConfigProperty<T>): T =
+    valueOrDefault(ruleSetConfigProperty.key, ruleSetConfigProperty.defaultValue)
```

Comment:
nit: This is probably a matter of taste, but I think I prefer the `value` function on the property instead of the config.

```suggestion
class RuleSetConfigProperty<T : Any>(private val key: String, private val defaultValue: T) {
    fun value(config: Config): T = config.valueOrDefault(key, defaultValue)
}
```

This would change the caller to
```kotlin
protected val isAndroid
    get() = FormattingProvider.android.value(ruleSetConfig)
```

Reply:
I have no preference so your advice rules.